src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -318,4 +318,66 @@ object messages {
         |$code2
         |""".stripMargin
   }
+
+  case class TopLevelImplicitClass(cdef: untpd.TypeDef)(implicit ctx: Context)
+    extends Message(10) {
+    val kind = "Syntax"
+
+    val msg = hl"""|An ${"implicit class"} may not be top-level"""
+
+    val explanation =
+    hl"""|implicit class are not allowed to be defined at the top-level.
+         |
+         |Declare ${cdef.name} inside of an ${"object"} or ${"package object"}
+         |then import the members of the object at the use site if needed, for example:
+         |
+         |package object Implicits {
+         |  implicit class ${cdef.name}...
+         |}
+         |
+         |import Implicits.${cdef.name}""".stripMargin
+  }
+
+  case class ImplicitCaseClass(cdef: untpd.TypeDef)(implicit ctx: Context)
+    extends Message(11) {
+    val kind = "Syntax"
+
+    val msg = hl"""|A ${"case class"} may not be defined as ${"implicit"}"""
+
+    val explanation =
+      hl"""|implicit classes may not be case classes. Instead use a plain class:
+           |
+           |implicit class ${cdef.name}...""".stripMargin
```

Comment:
If you want to, I mean I'd at least shorten it to not include the body of the definition (i.e. `rhs`). I would say that a similar method to what was described bellow would work for this too, and as such you can factor it out to a common method.

It is also possible to have a look at `dottydoc/src/.../DocASTPhase.scala` to get a sense of how the new doctool does it. Bear in mind that the doctool is working with a typed tree..